### PR TITLE
feat(tree): implement passive allocation order tracking

### PR DESCRIPTION
**Pull request by polaz**
## Summary
- Add order-aware query methods to buildStore for tracking passive allocation order
- Implement smart reordering with automatic dependency adjustment
- Add respec operations to remove passives from a given point onward
- Add order validation to ensure tree connectivity is maintained

## Changes
- **buildStore**: New methods for order queries (`getPassiveIndex`, `getAllocatedAtLevel`, etc.), reordering (`reorderPassive`, `insertPassiveAt`), and respec (`respecFromIndex`, `respecFromNode`)
- **PathFinder**: New utilities for order validation (`validateAllocationOrder`), dependency tracking (`getRequiredDependencies`, `getDependentNodes`), and smart reordering (`smartReorderNode`)
- **useTreeAllocation**: Exposed new order tracking methods for UI components

## Test plan
- [x] Unit tests for order-aware query methods (getPassiveIndex, getAllocatedUpToIndex, etc.)
- [x] Unit tests for respec operations (respecFromIndex, respecFromNode)
- [x] Unit tests for allocation order preservation
- [x] Unit tests for order validation (validateAllocationOrder)
- [x] Unit tests for dependency queries (getRequiredDependencies, getDependentNodes)
- [x] All existing tests pass (2063 tests)
- [x] Build succeeds

Closes https://github.com/polaz/pob2-web/issues/70

### CODE REVIEW

File: src/engine/tree/PathFinder.ts [hunk 1]
```diff
@@ -416,3 +416,287 @@ export function getRefundableNodes(
...
+    const nodesToMove = new Set([nodeId, ...missingDeps]);
+
+    for (let i = 0; i < currentOrder.length; i++) {
+      const id = currentOrder[i]!;
+
+      if (i === clampedIndex) {
+        // Insert moved nodes at target position
+        // First insert missing deps in their original relative order
+        for (const dep of missingDeps) {
+          if (currentOrder.indexOf(dep) > clampedIndex) {
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/engine/tree/PathFinder.ts]: Fixed: Added `buildIndexMap()` helper that creates an index map once at the start of `smartReorderNode`. All position lookups now use this map for O(1) access instead of repeated indexOf calls.

File: src/engine/tree/PathFinder.ts [hunk 2]
```diff
@@ -416,3 +416,287 @@ export function getRefundableNodes(
...
+        movedNodes.push(nodeId);
+        for (const dep of dependents) {
+          if (currentOrder.indexOf(dep) < clampedIndex) {
+            newOrder.push(dep);
+            movedNodes.push(dep);
+          }
+        }
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/engine/tree/PathFinder.ts]: Fixed: Extracted the reordering logic into two well-documented helper functions: `handleMoveEarlier()` and `handleMoveLater()`. Each function handles one direction of movement with clear parameter documentation.

File: test/unit/engine/tree/PathFinder.test.ts [hunk 3]
```diff
@@ -473,4 +477,153 @@ describe('PathFinder', () => {
...
+      expect(result.reason).toBe('Node is not in the allocation order');
+    });
+
+    it('handles moving within valid range', () => {
+      const order = ['2', '3', '4', '5'];
+      // Move '4' from position 2 to position 1
+      // This should work in linear tree since 4 depends on 3 which is at position 1
+      const result = smartReorderNode('4', 2, order, classStartId, nodes);
+      expect(result.success).toBe(true);
+    });
```
[the remark this answers is not in the document]
Comment (polaz) [reply test/unit/engine/tree/PathFinder.test.ts]: Fixed: Rewrote the test to validate actual reordering. Now tests moving node 3 to position 2 in a branching tree, which validates the smart reordering logic for independent branches.

File: src/engine/tree/PathFinder.ts [hunk 4]
```diff
@@ -416,3 +416,356 @@ export function getRefundableNodes(
...
+      for (const dep of dependents) {
+        const depIndex = indexMap.get(dep) ?? -1;
+        if (depIndex < clampedIndex) {
+          newOrder.push(dep);
+          movedNodes.push(dep);
+        }
+      }
+    }
+  }
+
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/engine/tree/PathFinder.ts]: Fixed: Refactored `handleMoveLater` to use a clearer approach with an `inserted` flag and `insertMovedNodes()` helper function. The logic now separates node skipping from insertion clearly.

File: src/engine/tree/PathFinder.ts [hunk 5]
```diff
@@ -416,3 +416,356 @@ export function getRefundableNodes(
...
+    // Skip nodes that were moved, unless they were already before target
+    if (!nodesToMove.has(id) || idIndex <= clampedIndex) {
+      if (!nodesToMove.has(id)) {
+        newOrder.push(id);
+      } else if (id !== nodeId) {
+        // Dep already before target position, keep it
+        newOrder.push(id);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/engine/tree/PathFinder.ts]: Fixed: Refactored `handleMoveEarlier` with clearer structure. Now explicitly builds `depsToMove` and `depsToMoveSet` upfront, then uses simple continue statements to skip moved nodes. Much easier to follow.

File: test/unit/engine/tree/PathFinder.test.ts [hunk 6]
```diff
@@ -473,4 +477,157 @@ describe('PathFinder', () => {
...
+      // In branching tree: 1-2-3 and 1-2-4-5
+      // Order: allocate 2, then branch to 3, then go to 4
+      const order = ['2', '3', '4'];
+      // Move '3' from index 1 to index 2 (after '4')
+      // This should succeed since '3' and '4' are independent branches from '2'
+      const result = smartReorderNode('3', 2, order, classStartId, nodes);
+      expect(result.success).toBe(true);
+      expect(result.movedNodes).toContain('3');
+    });
+  });
```
[the remark this answers is not in the document]
Comment (polaz) [reply test/unit/engine/tree/PathFinder.test.ts]: Fixed: Added comprehensive tests for smartReorderNode including moving to index 0, moving to end, and validation of returned order. Also added edge case tests for branching tree scenarios.

File: src/stores/buildStore.ts [hunk 7]
```diff
@@ -148,6 +235,147 @@ export const useBuildStore = defineStore('build', () => {
...
+      nodes
+    );
+
+    if (result.success) {
+      currentBuild.value.allocatedNodeIds = result.newOrder;
+      isDirty.value = true;
+    }
+
+    return result;
+  }
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/stores/buildStore.ts]: The `reorderPassive` method in buildStore is a thin wrapper that delegates to `smartReorderNode` (which is thoroughly tested) and then updates state. The core logic is tested in PathFinder.test.ts. Adding wrapper tests would duplicate coverage without adding value.

File: src/stores/buildStore.ts [hunk 8]
```diff
@@ -148,6 +235,147 @@ export const useBuildStore = defineStore('build', () => {
...
+  function validateOrder(
+    classStartId: string,
+    nodes: Map<string, PassiveNode>
+  ): OrderValidation {
+    return validateAllocationOrder(
+      currentBuild.value.allocatedNodeIds,
+      classStartId,
+      nodes
+    );
+  }
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/stores/buildStore.ts]: The `validateOrder` method is a thin wrapper around `validateAllocationOrder` which is tested in PathFinder.test.ts. The wrapper just passes through the build state to the tested function.

File: src/engine/tree/PathFinder.ts [hunk 9]
```diff
@@ -416,3 +416,388 @@ export function getRefundableNodes(
...
+      // Insert moved nodes at target position
+      // First insert deps (only those after target) in original relative order
+      for (const dep of depsToMove) {
+        newOrder.push(dep);
+        movedNodes.push(dep);
+      }
+      // Then insert the target node
+      newOrder.push(nodeId);
+      movedNodes.push(nodeId);
+    }
```
[the remark this answers is not in the document]
Comment (polaz) [reply src/engine/tree/PathFinder.ts]: Fixed: Added clarifying comment explaining the insertion timing and how the loop handles the target node position.

File: test/unit/engine/tree/PathFinder.test.ts [hunk 10]
```diff
@@ -473,4 +477,193 @@ describe('PathFinder', () => {
...
+      const result = getRequiredDependencies('2', classStartId, nodes);
+      expect(result).toEqual([]);
+    });
+
+    it('returns intermediate nodes for distant node', () => {
+      const result = getRequiredDependencies('5', classStartId, nodes);
+      // Path 5->4->3->2->1, dependencies are 2,3,4 (in order from start)
+      expect(result).toEqual(['2', '3', '4']);
+    });
+  });
```
[the remark this answers is not in the document]
Comment (polaz) [reply test/unit/engine/tree/PathFinder.test.ts]: Fixed: Added test case for unreachable node that verifies empty array is returned.

File: test/unit/engine/tree/PathFinder.test.ts [hunk 11]
```diff
@@ -473,4 +477,193 @@ describe('PathFinder', () => {
...
+
+    it('handles moving to end of order', () => {
+      nodes = buildBranchingTree();
+      const order = ['2', '3', '4'];
+      // Move '2' to the end - this should move dependents too
+      const result = smartReorderNode('2', 2, order, classStartId, nodes);
+      // '2' has dependents '3' and '4', so moving it later may fail
+      // or succeed with adjusted order
+      expect(typeof result.success).toBe('boolean');
+    });
```
[the remark this answers is not in the document]
Comment (polaz) [reply test/unit/engine/tree/PathFinder.test.ts]: Strengthened the test assertions. Now verifying `result.newOrder` contains the expected nodes and `result.movedNodes` includes the correct dependency/dependent nodes.

File: test/unit/engine/tree/PathFinder.test.ts [hunk 12]
```diff
@@ -473,4 +477,219 @@ describe('PathFinder', () => {
...
+
+    it('returns empty for leaf node', () => {
+      const allocated = new Set(['1', '2', '3']);
+      const result = getDependentNodes('3', allocated, classStartId, nodes);
+      expect(result.size).toBe(0);
+    });
+
+    it('returns dependent nodes for middle node', () => {
+      const allocated = new Set(['1', '2', '3', '4', '5']);
```
[the remark this answers is not in the document]
Comment (polaz) [reply test/unit/engine/tree/PathFinder.test.ts]: Fixed. Removed class start `1` from allocated set and added comment clarifying that class start is passed separately via the `classStartId` parameter.

File: test/unit/engine/tree/PathFinder.test.ts [hunk 13]
```diff
@@ -473,4 +477,219 @@ describe('PathFinder', () => {
...
+      const result = getDependentNodes('3', allocated, classStartId, nodes);
+      expect(result).toContain('4');
+      expect(result).toContain('5');
+    });
```
[the remark this answers is not in the document]
Comment (polaz) [reply test/unit/engine/tree/PathFinder.test.ts]: Fixed. Same change as above - removed `1` from allocated set and added clarifying comment.

File: test/unit/stores/buildStore.test.ts [hunk 14]
```diff
@@ -288,4 +288,196 @@ describe('buildStore', () => {
...
+
+    it('getPassiveLevel should return estimated level', () => {
+      const store = useBuildStore();
+      store.allocateNode('a');
+      store.allocateNode('b');
+      store.allocateNode('c');
+      store.allocateNode('d');
+
+      // PoE2: Level 1 gives 2 quest points (Act 1), so first 2 allocations are level 1
+      // Level 2 gives 1 more point (level-1=1), so 3rd allocation needs level 2
```
[the remark this answers is not in the document]
Comment (polaz) [reply test/unit/stores/buildStore.test.ts]: Fixed. Updated comment to clarify that these are points *spent* (allocations at indices), not points available.
